### ISSUE TITLE
OTB_PatchesExtraction reject all the shapefile points for SAR image and the model training fail

### ISSUE THREAD

**Issue by nitcheulegrand** [remicres/otbtf_tutorials#4]
Hello,
I try to modify the `create_model1.py` source code to traine the model model using patches extracted from Radar image with 2 polarities. My normalisation is made using two images (vv and vh polarities).  I run `PatchesExtraction` to extract patches of size 16x16. I then so expect to have as input a set of data of 16x16x2.
I have 7 classes in my model identified by numbers from 0 to 6. And my shapfile content 2 fields (id and class) both integer.
This is the folowing code I have.

```py
from tricks import *
import sys
import os

nclasses=7

def myModel(x):
  
  # input patches: 16x16x2
  conv1 = tf.layers.conv2d(inputs=x, filters=16, kernel_size=[5,5], padding="valid", 
                           activation=tf.nn.relu) # out size: 12x12x16
  pool1 = tf.layers.max_pooling2d(inputs=conv1, pool_size=[2, 2], strides=2) # out: 6x6x16
  conv2 = tf.layers.conv2d(inputs=pool1, filters=16, kernel_size=[3,3], padding="valid", 
                           activation=tf.nn.relu) # out size: 4x4x16
  pool2 = tf.layers.max_pooling2d(inputs=conv2, pool_size=[2, 2], strides=2) # out: 2x2x16
  conv3 = tf.layers.conv2d(inputs=pool2, filters=32, kernel_size=[2,2], padding="valid",
                           activation=tf.nn.relu) # out size: 1x1x32
  
  # Features
  features = tf.reshape(conv3, shape=[-1, 32], name="features")
  
  # Neurons for classes
  estimated = tf.layers.dense(inputs=features, units=nclasses, activation=None)
  estimated_label = tf.argmax(estimated, 1, name="prediction")

  return estimated, estimated_label
 
""" Main """
if len(sys.argv) != 2:
  print("Usage : <output directory for SavedModel>")
  sys.exit(1)

# Create the TensorFlow graph
with tf.Graph().as_default():
  
  # Placeholders
  x = tf.placeholder(tf.float32, [None, None, None, 2], name="x")
  y = tf.placeholder(tf.int32  , [None, None, None, 1], name="y")
  lr = tf.placeholder_with_default(tf.constant(0.0002, dtype=tf.float32, shape=[]),
                                   shape=[], name="lr")
  
  # Output
  y_estimated, y_label = myModel(x)
  
  # Loss function
  cost = tf.losses.sparse_softmax_cross_entropy(labels=tf.reshape(y, [-1, 1]), 
                                                logits=tf.reshape(y_estimated, [-1, nclasses]))
  
  # Optimizer
  optimizer = tf.train.AdamOptimizer(learning_rate=lr, name="optimizer").minimize(cost)
  
  # Initializer, saver, session
  init = tf.global_variables_initializer()
  saver = tf.train.Saver( max_to_keep=20 )
  sess = tf.Session()
  sess.run(init)

  # Create a SavedModel
  CreateSavedModel(sess, ["x:0", "y:0"], ["features:0", "prediction:0"], sys.argv[1])
```
When I run the script, the model is created and saved. But when I run the training command, it sames like it don't detect all the classes.
The kapa index is 0 or 1 and the confusion matrix is empty or is a 1x1 matrix.

Can you please help me and direct me about what to do.
Thanks in advance for your avaibility.

I relate to this some screenshot to better understand the behavior.

![Capture d’écran de 2021-08-09 19-56-34](https://user-images.githubusercontent.com/42095614/129490515-a80eba4b-e7d5-4b15-b5ac-bc065a463dc6.png)
![Capture d’écran de 2021-08-09 19-58-45](https://user-images.githubusercontent.com/42095614/129490520-d422e493-22d0-4951-8adf-11bd3bb134d9.png)
![Capture d’écran de 2021-08-09 19-59-08](https://user-images.githubusercontent.com/42095614/129490522-5989e838-8370-4a34-bbb8-b37e40dc29d8.png)
![Capture d’écran de 2021-08-09 19-59-32](https://user-images.githubusercontent.com/42095614/129490523-ab1ba44a-ba0f-4d6e-95d6-299d5bf8a569.png)
![Capture d’écran de 2021-08-09 20-00-16](https://user-images.githubusercontent.com/42095614/129490524-f934394a-80b7-471f-8092-e0ddf702ad93.png)
![Capture d’écran de 2021-08-09 20-01-54](https://user-images.githubusercontent.com/42095614/129490525-7f5ef9c3-af00-4eea-8118-a0399db4960d.png)
![Capture d’écran de 2021-08-09 20-09-56](https://user-images.githubusercontent.com/42095614/129490528-f0d881ca-2bb5-40bb-9ffa-58c7ba1b1d63.png)
![Capture d’écran de 2021-08-09 20-10-02](https://user-images.githubusercontent.com/42095614/129490531-e9f7d8d3-f169-418c-9be9-52b22c920aec.png)

**Comment by adminmyw** [comment]
Hi！I'm a student from China, I wanna learn something about deep-learning, will you contact my email 'mywswfu@163.com', THANK YOU VERY MUCH

**Comment by remicres** [comment]
Hi @nitcheulegrand 

It looks like the issue is with your data. Can you check that your labels?

What is the output of the following?
```
python -c "import gdal, numpy; d=gdal.Open('processing/patches/training/outlabel.tif'); print(numpy.histogram(d.GetRasterBand(1).ReadAsArray(), bins=range(7)))"
```

**Comment by nitcheulegrand** [comment]
Hi @remicres 
Thank for response.
I run the command an this is the output
`(array([175,   0,   0,   0,   0,   0]), array([0, 1, 2, 3, 4, 5, 6]))`
I my self think about the issue in data. I create a new shapefile in which I reduce points and let good distance between points but the result is the same.

**Comment by remicres** [comment]
The output shows that there are 175 samples of class `0` in labels, and no sample for the other classes.

You must carefully check the previous steps, when creating the labels. Check the vector layer, the class field, and how the polygons fit in the image, i.e. if it does not fall outside its extend or in no-data areas.

**Comment by nitcheulegrand** [comment]
My terrain truth is a shapefile made with points. The points where placed at the good place. It's not possible for the to be in the no-data areas or outside.
When I run SampleSelection the statistiques are good.
`on -in processing/normalized/normalized-image.tif -vec training/t-100.shp -field class -instats processing/sample/training/statistics.xml -out processing/sample/training/sampled-training-set-ponts.shp
`
This is the output
```
0       25      25      1
1       25      25      1
2       25      25      1
3       25      25      1
4       25      25      1
5       25      25      1
6       25      25      1
```

**Comment by nitcheulegrand** [comment]
When I run this command
`
python -c "import gdal, numpy; d=gdal.Open('processing/patches/training/outlabel.tif'); print(numpy.histogram(d.GetRasterBand(1).ReadAsArray(), bins=range(7)))"
`
It output an array of size 6 and an array of size 7. I expected but array to be of size 7.

**Comment by nitcheulegrand** [comment]
Oh!
I can see in the PatchesExtraction output that there is something wrong.
```
2021-08-18 12:33:25 (INFO) PatchesExtraction: Number of samples collected: 0
2021-08-18 12:33:25 (INFO) PatchesExtraction: Number of samples rejected : 175
```
This is the full output
```
2021-08-18 12:33:25 (INFO) PatchesExtraction: Default RAM limit for OTB is 256 MB
2021-08-18 12:33:25 (INFO) PatchesExtraction: GDAL maximum cache size is 197 MB
2021-08-18 12:33:25 (INFO) PatchesExtraction: OTB will use at most 2 threads
Sampling patches: 100% [**************************************************]
2021-08-18 12:33:25 (INFO) PatchesExtraction: Number of samples collected: 0
2021-08-18 12:33:25 (INFO) PatchesExtraction: Number of samples rejected : 175
2021-08-18 12:33:25 (INFO): File processing/patches/training/outpatches_16x16.tif will be written in 1 blocks of 16x2800 pixels
Writing processing/patches/training/outpatches_16x16.tif...: 0% [                                                  ]
2021-08-18 12:33:25 (WARNING): Could not get the source process object. Progress report might be buggy
Writing processing/patches/training/outpatches_16x16.tif...: 100% [**************************************************] (0s)
2021-08-18 12:33:25 (INFO): Estimated memory for full processing: 0.00224113MB (avail.: 256 MB), optimal image partitioning: 1 blocks
2021-08-18 12:33:25 (INFO): File processing/patches/training/outlabel.tif will be written in 1 blocks of 1x175 pixels
Writing processing/patches/training/outlabel.tif...: 100% [**************************************************] (0s)

```
How ca I resolve this.
The PatchesExtraction reject the sampled data.

**Comment by remicres** [comment]
Hi, @nitcheulegrand 

The **rejected samples** are not kept because either: 
- falling outside the image boundaries, or
- contains at least one no-data pixel (if you have specified a no-data value in `PatchesExtraction`)

Maybe the projection is wrong? Did you try to project your vector data in the same projection as your image?

**Comment by nitcheulegrand** [comment]
The problem was caused by the projection but I can explain how.
To solve the problem, I used GDAL and extracted a portion of the image using the same projection and I worked instead with the portion.
I ran the `PatchesExtraction` on the extracted image and it worked well.
Thanks for your assistance.
